src/node.h

```diff
@@ -731,12 +731,27 @@ struct StartExecutionCallbackInfo {
 
 using StartExecutionCallback =
     std::function<v8::MaybeLocal<v8::Value>(const StartExecutionCallbackInfo&)>;
+using EmbedderPreloadCallback =
+    std::function<void(Environment* env,
+                       v8::Local<v8::Value> process,
+                       v8::Local<v8::Value> require)>;
 
+// Run initialization for the environment.
+//
+// The |preload| function will run before executing the entry point, which
+// is usually used by embedders to inject scripts. The function is executed
+// with preload(process, require), and the passed require function has access
+// to internal Node.js modules. The |preload| function is inherited by worker
+// threads and thus will run in work threads, so make sure the function is
+// thread-safe.
 NODE_EXTERN v8::MaybeLocal<v8::Value> LoadEnvironment(
```

Comment:
It maybe useful to mention two things about this callback:

1. There is no stability guarantee about the internals exposed to the internal require. Expect breakages whenever the embedder updates the Node.js version that they embed (actually if access to internals is not strictly necessary, I'd prefer to have a special wrapper like the one we have for mksnapshot to filter out internals, or make the access to internals behind some flags)
2. The callback is run before the user land module loader is ready to run any user code so it is safe to assume that at this point, no user code has been run yet.

Reply:
I have updated the comments.

Electron does need to import the internal modules:
https://github.com/electron/electron/blob/12d7a8ff6605cca218e3e0a507de9a8f530c2abd/lib/node/asar-fs-wrapper.ts#L49-L52

The preload scripts of Electron apps are loaded without access to internal modules though:
https://github.com/electron/electron/blob/2495207a2fe158ace05eae2cbbcd2e8679d89c19/lib/node/init.ts#L13-L27